compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs

```diff
@@ -0,0 +1,71 @@
+-- Copyright (c) 2020 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.Compiler.Validate (validateDar) where
+
+import Control.Monad (forM_)
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Proto3.Archive as Archive
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.NameMap as NM
+
+import DA.Daml.Compiler.ExtractDar (extractDar,ExtractedDar(..))
+import DA.Daml.LF.Ast.Version
+import DA.Daml.LF.Ast.World (initWorld)
+import DA.Pretty (renderPretty)
+import qualified DA.Daml.LF.TypeChecker as TC (checkModule)
+import qualified DA.Daml.LF.TypeChecker.Error as TC
+
+-- | Monad for performing Validation
+newtype Validation a = Validation (Either ValidationError a) deriving (Functor,Applicative,Monad)
+
+data ValidationError
+  = VeArchiveError FilePath Archive.ArchiveError
+  | VeTypeError TC.Error
+
+validationError :: ValidationError -> Validation a
+validationError = Validation . Left
+
+-- | Run a Validation, throwing an error when the validation fails
+runValidation :: Validation a -> a
+runValidation = \case
+  Validation (Left e) -> error (show e)
+  Validation (Right a) -> a
+
+instance Show ValidationError where
+  show = \case
+    VeArchiveError fp err -> unlines
+      [ "Invalid DAR."
+      , "DALF entry cannot be decoded: " <> fp
+      , show err ]
+    VeTypeError err -> unlines
+      [ "Invalid DAR."
+      , "The DAR is not well typed."
+      , renderPretty err ]
+
+-- | Validate a loaded DAR: that all DALFs are well-typed and consequently that the DAR is closed
+validateDar :: FilePath -> IO Int
+validateDar inFile = do
+  ExtractedDar{edDalfs} <- extractDar inFile
+  return $ runValidation $ do
+    xps <- mapM (validateEntry Archive.DecodeAsDependency) edDalfs
```

Comment:
I was trying to see if laziness kicks in here (looks like it doesn’t and it’s forced somehow), however there is a much bigger issue:

`damlc validate-dar` is unable to validate a DAR created with `daml new foobar && cd foobar && daml build`:

```
damlc: Invalid DAR.
The DAR is not well typed.
error type checking template Main.Asset precondition:
  type mismatch:
  * expected type:
      c38eb615743a614183ee4a655c2a31dc732269ed50e0c54894d3eab1ade6cb80:Main:Asset
  * found type:
      Main:Asset
  * expression:
      this

CallStack (from HasCallStack):
  error, called at compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs:33:26 in compilerZSdamlcZSdaml-compilerZSdaml-compiler:DA.Daml.Compiler.Validate
```

It looks like this is probably caused by the fact that you use `initWorld` which only initializes dependency packages instead of  `initWorldSelf` which also intiializes the current package which is necessary to resolve `PRSelf`. This is probably coming from https://github.com/digital-asset/daml/blob/master/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs#L647.

I believe any code that uses a template will exhibit this so making a testcase should be fairly straightforward hopefully.

Reply:
This is now fixed.